rust/agama-lib/src/product/store.rs

```diff
@@ -0,0 +1,63 @@
+//! Implements the store for the product settings.
+
+use super::{ProductClient, ProductSettings};
+use crate::error::ServiceError;
+use crate::manager::ManagerClient;
+use zbus::Connection;
+
+/// Loads and stores the product settings from/to the D-Bus service.
+pub struct ProductStore<'a> {
+    product_client: ProductClient<'a>,
+    manager_client: ManagerClient<'a>,
+}
+
+impl<'a> ProductStore<'a> {
+    pub async fn new(connection: Connection) -> Result<ProductStore<'a>, ServiceError> {
+        Ok(Self {
+            product_client: ProductClient::new(connection.clone()).await?,
+            manager_client: ManagerClient::new(connection).await?,
+        })
+    }
+
+    pub async fn load(&self) -> Result<ProductSettings, ServiceError> {
+        let product = self.product_client.product().await?;
+        let registration_code = self.product_client.registration_code().await?;
+        let email = self.product_client.email().await?;
+
+        Ok(ProductSettings {
+            id: Some(product),
+            registration_code: Some(registration_code),
+            registration_email: Some(email),
+        })
+    }
+
+    pub async fn store(&self, settings: &ProductSettings) -> Result<(), ServiceError> {
+        let mut probe = false;
+        if let Some(product) = &settings.id {
+            let existing_product = self.product_client.product().await?;
+            if *product != existing_product {
+                // avoid selecting same product and unnecessary probe
+                self.product_client.select_product(product).await?;
+                probe = true;
+            }
+        }
+        if let Some(reg_code) = &settings.registration_code {
+            let (result, message);
+            if let Some(email) = &settings.registration_email {
+                (result, message) = self.product_client.register(reg_code, email).await?;
+            } else {
+                (result, message) = self.product_client.register(reg_code, "").await?;
+            }
+            if result != 0 {
+                return Err(ServiceError::FailedRegistration(message));
```

Comment:
Thanks!

... ... but ... my testing shows:
```console
# agama config load agama-lib/share/examples/profile_Dolomite.json
Registration failed: 'Connection to registration server failed'
```

So far so good, right?

```console
# tail -n2 daemon.log
Announcing system to https://scc.suse.com ...
[ERROR]: software: Error connecting to registration server: Unknown Registration Code.
```

So the CLI would have me debug network connections instead of checking for typos in the regcode. We still need to improve this.

Reply:
When I quickly insert the integer result code into `FailedRegistration` it is revealed to be `6` which means "API error" which is a disappointment, I hoped to see `8` "incorrect credentials".

Reply:
aaand I have fixed it in our Ruby backend :smile: